### PR TITLE
ui: local forage implementation

### PR DESCRIPTION
**Pull request by karolina-siemieniuk-morawska**
* Use localForage instead of localStorage in the app

* ref: #1359

changed dealing with async value

### CODE REVIEW

File: ui/src/common/storage.js [hunk 1]
```diff
@@ -1,9 +1,25 @@
+import localforage from 'localforage';
+
+localforage.config({
+  name: 'inspirehep',
+  driver: localforage.INDEXEDDB,
```
Comment (drjova) [inline ui/src/common/storage.js:5]:
```suggestion
  driver:[localforage.INDEXEDDB, localforage.LOCALSTORAGE)]
```

File: ui/src/common/storage.js [hunk 2]
```diff
@@ -15,7 +31,8 @@ function isStorageAvailable() {
         e.code === 1014 ||
         e.name === 'QuotaExceededError' ||
         e.name === 'NS_ERROR_DOM_QUOTA_REACHED') &&
-      (storage && storage.length !== 0)
+      storage &&
+      storage.length !== 0
```
Comment (drjova) [inline ui/src/common/storage.js:35]: why you removed the parenthesis?

File: ui/src/common/storage.js [hunk 3]
```diff
@@ -26,8 +43,8 @@ const storage = {
   getSync(key) {
     // eslint-disable-next-line react/no-this-in-sfc
     if (this.isAvailable) {
-      const rawValue = localStorage.getItem(key);
-      return JSON.parse(rawValue);
+      const rawValue = localforage.getItem(key).then(item => JSON.parse(item));
+      return rawValue;
```
Comment (drjova) [inline ui/src/common/storage.js:47]: This will return undefined rawValue 🤔 why not use async?

Comment (drjova) [reply ui/src/common/storage.js:47]: Also according to the documentation:

> [...] localForage automatically does JSON.parse() and JSON.stringify() when getting/setting values.

File: ui/src/common/storage.js (hunk 2 above)
Comment (karolina-siemieniuk-morawska) [reply ui/src/common/storage.js:35]: I believe it was prettier who did this or it must have been by accident, I'll revert this change

File: ui/src/common/storage.js (hunk 3 above)
Comment (karolina-siemieniuk-morawska) [reply ui/src/common/storage.js]: Changed to use asynchronous syntax and removed json parsing

File: ui/src/common/storage.js (hunk 1 above)
Comment (karolina-siemieniuk-morawska) [reply ui/src/common/storage.js]: added fallback driver

File: ui/src/common/storage.js [hunk 4]
```diff
@@ -1,9 +1,25 @@
...
+});
+
+// to not lose user settings saved in localStorage rewrite them to localforage
+for (let i = 0; i < localStorage.length; i += 1) {
+  const key = localStorage.key(i);
+  const value = localStorage.getItem(key);
+
+  localforage.setItem(key, value);
+}
+
```
Comment (drjova) [inline ui/src/common/storage.js:16]: Last thing I promise, could you please add this in a try catch?

Comment (karolina-siemieniuk-morawska) [reply ui/src/common/storage.js]: Done 👾

File: ui/src/common/storage.js [hunk 5]
```diff
@@ -1,12 +1,30 @@
...
   try {
     // eslint-disable-next-line dot-notation
-    storage = window['localStorage'];
+    storage = localforage;
     const test = '__test__';
-    storage.setItem(test, test);
-    storage.removeItem(test);
+    await storage.setItem(test, test);
+    await storage.removeItem(test);
+    rewriteUserSettingsFromLocalStorage();
```
Comment (drjova) [inline ui/src/common/storage.js:27]: shouldn't this be async as well?

**Review by drjova — COMMENTED** [review]
LGTM just the comment and we can test it on QA :)

**Review by drjova — APPROVED** [review]

**Pull request merged**